### ISSUE TITLE
json_prepare_post callback not triggered on parent after featured image added

### ISSUE THREAD

**Issue by iandunn** [WP-API/WP-API#1090]
If I register a callback for `json_prepare_post` at priority `9` and then request `/posts/{id}` on a post that has a featured image, the function will be called on the parent post and then again on the attachment. That's the expected behavior.

But then, if I change the priority of the callback to `11` (after `WP_JSON_Media::add_thumbnail_data()` runs), then the function will only be called once, on the attachment; it won't be called on the post itself like it should be.

If I request a post without a featured image, then the function will be called regardless of the priority.

So if a post has a featured image, any functions hooked in after `add_thumbnail_data()` is called will be ignored.

This example below should always result in an error being returned to the request, but if you request a post with a featured image then you'll get the normal post object instead.

```
function wp_api_1090( $prepared_post, $raw_post, $context ) {
	return new WP_Error( '1090', "You should always get this message, but you won't if the post has a featured image." );
}
add_filter( 'json_prepare_post', 'wp_api_1090', 11, 3 );
```

**Comment by iandunn** [comment]
Sounds like this might be a result of [WP #17817](https://core.trac.wordpress.org/ticket/17817).

**Comment by iandunn** [comment]
Yeah, I think it's caused by [WP #17817](https://core.trac.wordpress.org/ticket/17817), which sounds like it might be fixed in 4.3.

In the meantime, I don't need featured images, so I'm just unhooking `add_thumbnail_data()` as a workaround:

```
/*
 * Avoid nested callback conflicts by de-registering WP_JSON_Media::add_thumbnail_data().
 *
 * There's a Core bug (#17817) where nested callbacks conflict with each other. If a post has a featured image,
 * then `json_prepare_post` will be called recursively, and `wcorg_expose_whitelisted_json_api_meta_data()` will
 * be short-circuited.
 *
 * Hopefully the fix for #17817 will land in 4.3, but until then we can avoid the issue by de-registering
 * `add_thumbnail_data()`. This obviously has the downside of removing featured image data from the API responses,
 * but we don't have a compelling use case for that data right now, and I don't see a better alternative. Ensuring
 * that only whitelisted data is exposed is a privacy issue, and therefore trumps pretty much everything else.
 *
 * This is fragile, and will probably break when WP-API's `develop` branch is merged into `master`, so -- like
 * everything else in this plugin -- it should be re-tested and updated when new versions of the API are
 * released.
 *
 * See https://github.com/WP-API/WP-API/issues/1090
 * See https://core.trac.wordpress.org/ticket/17817
 */
function wcorg_avoid_nested_callback_conflicts() {
	global $wp_json_media;

	remove_filter( 'json_prepare_post', array( $wp_json_media, 'add_thumbnail_data' ), 10, 3 );
}
add_action( 'wp_json_server_before_serve', 'wcorg_avoid_nested_callback_conflicts', 11 );    // after `json_api_default_filters()`
```
